File: ios/nodes/AudioBufferSourceNode/AudioBufferSourceNode.m
#import <AudioBufferSourceNode.h>
#import "Constants.h"

@implementation AudioBufferSourceNode

- (instancetype)initWithContext:(AudioContext *)context
{
  if (self != [super initWithContext:context]) {
    return self;
  }

    _loop = true;
  _isPlaying = NO;
    _bufferIndex = 0;
    self.channelCount = 2;
    _buffer = [[RNAA_AudioBuffer alloc] initWithSampleRate:context.sampleRate length:[Constants bufferSize] numberOfChannels:2];

  self.numberOfOutputs = 1;
  self.numberOfInputs = 0;

    [self initAudioSourceNode:_buffer];

  return self;
}

- (void)initAudioSourceNode:(RNAA_AudioBuffer *)buffer {
    _format = [[AVAudioFormat alloc] initStandardFormatWithSampleRate:_buffer.sampleRate channels:_buffer.numberOfChannels];
    __weak typeof(self) weakSelf = self;
    _sourceNode = [[AVAudioSourceNode alloc] initWithFormat:_format
                                                renderBlock:^OSStatus(
                                                    BOOL *isSilence,
                                                    const AudioTimeStamp *timestamp,
                                                    AVAudioFrameCount frameCount,
                                                    AudioBufferList *outputData) {
                                                  return [weakSelf renderCallbackWithIsSilence:isSilence
                                                                                     timestamp:timestamp
                                                                                    frameCount:frameCount
                                                                                    outputData:outputData];
                                                }];
    [self.context.audioEngine attachNode:self.sourceNode];
    [self.context.audioEngine connect:self.sourceNode to:self.context.audioEngine.mainMixerNode format:self.format];

    if (!self.context.audioEngine.isRunning) {
      NSError *error = nil;
      if (![self.context.audioEngine startAndReturnError:&error]) {
        NSLog(@"Error starting audio engine: %@", [error localizedDescription]);
      }
    }
}

- (void)cleanup
{
  if (_isPlaying) {
    [self stopPlayback];
  }

  [self.context.audioEngine detachNode:self.sourceNode];

  _format = nil;
}

- (bool)getLoop {
    return _loop;
}

- (void)setLoop:(bool)loop {
    _loop = loop;
}

- (void)setBuffer:(RNAA_AudioBuffer *)buffer {
    _buffer = buffer;
    [self initAudioSourceNode:_buffer];
}

- (void)start:(double)time
{
  if (_isPlaying) {
    return;
  }

  double delay = time - [self.context getCurrentTime];
  if (delay <= 0) {
    [self startPlayback];
  } else {
    [NSTimer scheduledTimerWithTimeInterval:delay
                                     target:self
                                   selector:@selector(startPlayback)
                                   userInfo:nil
                                    repeats:NO];
  }
}

- (void)startPlayback
{
  _isPlaying = YES;
}

- (void)stop:(double)time
{
  if (!_isPlaying) {
    return;
  }

  double delay = time - [self.context getCurrentTime];
  if (delay <= 0) {
    [self stopPlayback];
  } else {
    [NSTimer scheduledTimerWithTimeInterval:delay target:self selector:@selector(stopPlayback) userInfo:nil repeats:NO];
  }
}

- (void)stopPlayback
{
  _isPlaying = NO;
}

- (OSStatus)renderCallbackWithIsSilence:(BOOL *)isSilence
                              timestamp:(const AudioTimeStamp *)timestamp
                             frameCount:(AVAudioFrameCount)frameCount
                             outputData:(AudioBufferList *)outputData
{
  for (int frame = 0; frame < frameCount; frame += 1) {
    if (!_isPlaying) {
        for (int channel = 0; channel < _buffer.numberOfChannels; channel += 1) {
            outputData->mBuffers[channel].mData = 0;
        }
      continue;
    }
      
      for (int channel = 0; channel < _buffer.numberOfChannels; channel += 1) {
          outputData->mBuffers[channel].mData = [_buffer getChannelDataForChannel:channel][_bufferIndex];
      }
      
      _bufferIndex += 1;

      if (_bufferIndex >= _buffer.length) {
          _bufferIndex = 0;
      }
  }

  [self process:frameCount bufferList:outputData];

  return noErr;
}

- (Boolean)getIsPlaying
{
  return _isPlaying;
}

@end

File: packages/react-native-audio-api/android/src/main/cpp/AudioBufferSourceNode/AudioBufferSourceNode.cpp
#include "AudioBufferSourceNode.h"
#include "AudioContext.h"

namespace audioapi {

AudioBufferSourceNode::AudioBufferSourceNode(AudioContext *context)
    : AudioScheduledSourceNode(context), loop_(false), bufferIndex_(0) {
  numberOfInputs_ = 0;
  buffer_ = std::make_shared<AudioBuffer>(2, 428, context->getSampleRate());
}

bool AudioBufferSourceNode::getLoop() const {
  return loop_;
}

std::shared_ptr<AudioBuffer> AudioBufferSourceNode::getBuffer() const {
  return buffer_;
}

void AudioBufferSourceNode::setLoop(bool loop) {
  loop_ = loop;
}

void AudioBufferSourceNode::setBuffer(
    const std::shared_ptr<AudioBuffer> &buffer) {
  buffer_ = buffer->mix(channelCount_);
}

bool AudioBufferSourceNode::processAudio(float *audioData, int32_t numFrames) {
  if (!isPlaying_) {
    return false;
  } else {
    for (int i = 0; i < numFrames; ++i) {
      for (int j = 0; j < channelCount_; j++) {
        audioData[i * channelCount_ + j] =
            buffer_->getChannelData(j)[bufferIndex_];
      }

      bufferIndex_++;

      if (bufferIndex_ >= buffer_->getLength()) {
        if (loop_) {
          bufferIndex_ = 0;
        } else {
          isPlaying_ = false;
          break;
        }
      }
    }

    return true;
  }
}
} // namespace audioapi
